lighthouse-core/lib/network-request.js

```diff
@@ -0,0 +1,260 @@
+    this.initiatorRequest = () => /** @type {NetworkRequest|undefined} */ (undefined);
+    this._responseHeaders = /** @type {LH.WebInspector.HeaderValue[]} */ ([]);
+
+    this._fetchedViaServiceWorker = false;
+    this._frameId = /** @type {string|undefined} */ ('');
+    this._isLinkPreload = false;
+
+    // Make sure we're compatible with old WebInspector.NetworkRequest
+    // eslint-disable-next-line no-unused-vars
+    const record = /** @type {LH.WebInspector.NetworkRequest} */ (this);
+  }
+
+  /**
+   * @return {NetworkRequest}
+   */
+  clone() {
+    return Object.assign(new NetworkRequest(), this);
+  }
+
+  /**
+   * @param {NetworkRequest} initiator
+   */
+  setInitiatorRequest(initiator) {
+    this.initiatorRequest = () => initiator;
+  }
+
+  /**
+   * @param {LH.Crdp.Network.RequestWillBeSentEvent} data
+   */
+  onRequestWillBeSent(data) {
+    this.requestId = data.requestId;
+    this._requestId = data.requestId;
+
+    const url = new URL(data.request.url);
+    this.url = data.request.url;
+    this._url = data.request.url;
+    this.documentURL = data.documentURL;
+    this.parsedURL = {
+      scheme: url.protocol.split(':')[0],
+      // Intentional, DevTools uses different terminology
+      host: url.hostname,
+      securityOrigin: () => url.origin,
+    };
+    this.isSecure = SECURE_SCHEMES.includes(this.parsedURL.scheme);
+
+    this.startTime = data.timestamp;
+
+    this.requestMethod = data.request.method;
+
+    this._initiator = data.initiator;
+    this._resourceType = data.type && resourceTypes[data.type];
+    this.priority = () => data.request.initialPriority;
+
+    this._frameId = data.frameId;
+    this._isLinkPreload = data.initiator.type === 'preload' || !!data.request.isLinkPreload;
+  }
+
+  onRequestServedFromCache() {
+    this._fromMemoryCache = true;
+  }
+
+  /**
+   * @param {LH.Crdp.Network.ResponseReceivedEvent} data
+   */
+  onResponseReceived(data) {
+    this._onResponse(data.response, data.timestamp, data.type);
+    this._frameId = data.frameId;
+  }
+
+  /**
+   * @param {LH.Crdp.Network.DataReceivedEvent} data
+   */
+  onDataReceived(data) {
+    this._resourceSize += data.dataLength;
+    if (data.encodedDataLength !== -1) {
+      this.transferSize += data.encodedDataLength;
+    }
+  }
+
+  /**
+   * @param {LH.Crdp.Network.LoadingFinishedEvent} data
+   */
+  onLoadingFinished(data) {
+    if (this.finished) return;
+
+    this.finished = true;
+    this.endTime = data.timestamp;
+    if (data.encodedDataLength >= 0) {
+      this.transferSize = data.encodedDataLength;
+    }
+
+    this._updateResponseReceivedTimeIfNecessary();
+  }
+
+  /**
+   * @param {LH.Crdp.Network.LoadingFailedEvent} data
+   */
+  onLoadingFailed(data) {
+    if (this.finished) return;
+
+    this.finished = true;
+    this.endTime = data.timestamp;
+
+    this.failed = true;
+    this._resourceType = data.type && resourceTypes[data.type];
```

Comment:
why updated here?

Reply:
because DT does :) https://cs.chromium.org/chromium/src/third_party/blink/renderer/devtools/front_end/sdk/NetworkManager.js?type=cs&q=%22setResourceType%22+-f:out+lang:js&sq=package:chromium&g=0&l=600

I think it's because resourceTypes can always lie, so you can always get more information when the response comes back, also probably has to do with signed-exchanges and some corner cases we don't really care about

Reply:
ah, yeah, good to mirror the real deal. I was also wondering because `onLoadingFinished` doesn't do the same